keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,819 @@
+
+### Goals
+
+* Add field and label selectors to authorization attributes for List, Watch, and DeleteCollection verbs.
+* Add field and label selectors to webhook authorization types.
+* Add field and label selectors to SelfSubjectAccessReview (SSAR) and SubjectAccessReview (SAR).
+* Update node authorizer to restrict on nodeName field selector.
+
+### Non-Goals
+
+* Create a generic in-tree authorizer that manages field or label selectors.
+* Expand the audit surface area, since requestURI is already included
+* Expand the admission surface area (admission.Attributes, AdmissionReview, CEL authorizer implementation available to admission) 
+  since admission verbs don't support field/label selectors
+
+
+## Proposal
+
+List, Watch, and DeleteCollection requests directly have field and label selector options.
+A single-item List or Watch request is still a list as normal (including selectors), but also includes a name.
+
+### Authorization Attributes changes
+
+The authorization attributes have easy access to the query parameter field and label selectors.
+To avoid confusion, field and label selectors will not be included in authorization attributes for verbs where the field
+selector has no semantic meaning.
+In practice this means that (for now), only List, Watch, and DeleteCollection have field and label selectors.
+SubjectAccessReviews submitted to the kube-apiserver with verbs that do not honor the selectors will have the field and
+label selector attributes removed prior to checking authorization.
+Webhook authors: remember that the list of verbs accepting field and label selectors may change over time.
+If the kube-apiserver sends the FieldSelector or LabelSelector to a webhook, the kube-apiserver intends to honor the selector attributes.
+Any authorizer that gets an error from `ParseFieldSelector` or `ParseLabelSelector` may attempt to authorize without
+field or label selectors since that will authorize using a wider permission (field and label selectors can only reduce access).
+
+```go
+type Attributes interface {
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+	// It can return an error if the field selector cannot be parsed.
+    // Remember that field selector formats vary based on the version of the API being used!
+	ParseFieldSelector() ([]Requirement, error)
+  
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+    // It can return an error if the field selector cannot be parsed.
+    ParseLabelSelector() ([]Requirement, error)
+```
+
+#### Future-proofing your authentication webhook for future verbs
+
+As of 1.31, the only verbs with field and label selectors are List, Watch, and DeleteCollection.
+<<[UNRESOLVED deads2k, liggitt, jpbetz, sttts: this is a future promise for the kube-apiserver behavior if we add it ]>>
+In the future, the kube-apiserver may add field and label selectors to Create, Update, and Delete.
```

Comment:
If, in the future, we add field/label selectors to Create/Update/Patch/Delete, is get the only remaining unsupported CRUD operation? Should we support it in the future too?

Reply:
Thinking about this more? Is an explicit selector ever needed for single resource operations? Is the actual value of the labels and fields sufficient for any authz checks that would need to be made?

Reply:
authz doesn't have access to the body of the incoming object or the persisted content of the existing object

The query here would be like a self-attesting precondition the writer could optionally add that would be visible to authz and would have to be verified/blocking at the storage layer for the write to succeed.

The opt-in nature and the duplication with the body content is… a little weird, so I'm not sure if this is a direction we'd ever go, but defining how the precondition would be treated *if* we did is important.

Reply:
OK, this explains it. I was wondering if this might be the motivation.  Including this background in the KEP about _why_ we could consider passing a field selector alongside a single resource operation that redundantly states information already in the resource seems valuable here.